File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        ListNode dummyNode = new ListNode(0);
        dummyNode.next = head;
        ListNode leftNode = dummyNode;
        ListNode rightNode = dummyNode;
        ListNode lastNodeofReversedList = dummyNode;
        ListNode curr = dummyNode;
        int counter = 0;

        if (left == right) {
            return head;
        }

        while (curr != null) {
            if (counter == left - 1) {
                leftNode = curr;
            }

            ListNode prev = leftNode;
            while (counter >= left && counter <= right ) {
                if (counter == left) {
                    lastNodeofReversedList = curr;
                }

                if (counter == right) {
                    rightNode = curr.next;
                }
                ListNode temp = curr.next;
                curr.next = prev;
                prev = curr;
                if (counter == right) {
                    break;
                }
                curr = temp;
                counter++;
            }

            if (counter == right) {
                lastNodeofReversedList.next = rightNode;
                leftNode.next = prev;
            }
            counter++;
            curr = curr.next;
        }

        return dummyNode.next;
    }
}

File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if left == right:
            return head
        
        dummy = ListNode()
        dummy.next = head
        prev = dummy

        for _ in range(left - 1):
            prev = prev.next
        
        curr = prev.next
        then = curr.next
        for _ in range(right - left):
            curr.next = then.next
            then.next = prev.next
            prev.next = then
            then = curr.next
        
        return dummy.next